File: java/Basic/src/DSA/LinkedList/LinkedList.java
package DSA.LinkedList;

public class LinkedList {
    private Node head;
    private int size;

    public void append(Object data) {
        Node node = new Node(data);
        if (head == null) {
            head = node;
        } else {
            Node current = head;
            while (current.getNext() != null) {
                current = current.getNext();
            }
            current.setNext(node);
        }
        size += 1;
    }

    public void insert(Object data) {
        Node node = new Node(data);
        if (head == null) {
            head = node;
        } else {
            Node temp = head;
            head = node;
            head.setNext(temp);
        }
        size += 1;
    }

    public void removeHead() {
        head = head.getNext();
        size -= 1;
    }

    public void removeTail() {
        Node temp = head.getNext();

        while(temp.getNext() != null) {
            if (temp.getNext().getNext() == null) {
                temp.setNext(null);
                break;
            }
            temp = temp.getNext();

        }
        size -= 1;
    }

    public void remove(Object data) {
        if (head == null) {
            System.out.println("Linked List doesn't have data.");
        } else if (head.getNext() == null) {
            if (head.getData() == data) {
                head = null;
            } else {
                System.out.println("Data not founded.");
            }
        } else {
            Node prev = head;
            Node current = head;
            Node temp = null;
            boolean loop = true;

            while (loop) {
                if (current.getData() == data && current.getNext() != null) {
                    temp = current.getNext();
                    if (current == prev) {
                        head = temp;
                    } else {
                        prev.setNext(temp);
                    }
                    loop = false;
                } else if (current.getData() == data) {
                    temp = current;
                    prev.setNext(null);

                    loop = false;
                } else if (current.getNext() != null) {
                    if (prev == current) {
                        current = current.getNext();
                    } else {
                        current = current.getNext();
                        prev = prev.getNext();
                    }
                } else {
                    loop = false;
                }
            }

            if (temp != null) {
                System.out.printf("Node Data %s Berhasil Dihapus.\n", data);
                size -= 1;
            } else {
                System.out.println("Data not found.");
            }
        }
    }

    public void print() {
        StringBuilder text = new StringBuilder("Linked List Component = ");
        Node current = head;

        while(current.getNext() != null) {
            text.append(current.getData()).append(" -> ");
            current = current.getNext();
        }
        text.append(current.getData());
        System.out.println(text);
    }

    public void length() {
        System.out.printf("Jumlah Panjang data adalah %s\n", size);
    }
}

File: python/basic/DSA/LinkedList/LinkedList.py
class Node:
    def __init__(self, x):
        self.data: any = x
        self.next: Node = None
    
class LinkedList:
    def __init__(self):
        self.head: Node = None
        self.size: int = 0
    
    def append(self, data):
        newNode: Node = Node(data)

        if (self.head == None):
            self.head = newNode
        else:
            current = self.head
            while current.next != None:
                current = current.next
            current.next = newNode
        self.size += 1
    
    def insert(self, data):
        newNode: Node = Node(data)
        if self.head == None:
            self.head = newNode
        else:
            newNode.next = self.head
            self.head = newNode
        self.size += 1
    
    def deleteHead(self):
        if self.head != None:
            temp = self.head
            self.head = temp.next
            self.size -= 1
            print(temp.data)
        else:
            print("Delete Error, No data in LinkedList")
    
    def deleteTail(self):
        if self.head.next != None:
            if self.head.next == None:
                self.head = None
            else:
                temp = self.head
                while temp.next.next != None:
                    temp = temp.next
                lastNode: Node = temp.next
                temp.next = None
                print(lastNode.data)
                lastNode = None
                self.size -= 1
    
    def deleteNode(self, data):
        if self.head == None:
            print("Linked List belum memiliki data.")
        elif self.head.next == None:
            if self.head.data == data:
                self.head = None
            else:
                print("Node Tidak ditemukan.")
        else:
            prevNode: Node = self.head
            current: Node = self.head
            temp: Node = None

            loop = True
            while loop:
                if current.data == data and current.next:
                    temp = current.next
                    loop = False
                elif current.data == data:
                    temp = current
                    loop = False
                elif current.next != None:
                    if prevNode == current:
                        current = current.next
                    else:
                        prevNode = prevNode.next
                        current = current.next
                else:
                    loop = False

            if temp and current == prevNode:
                self.head = temp
                print(f"Node Data {data} Berhasil Dihapus.")
            elif temp == current:
                prevNode.next = None
                print(f"Node Data {data} Berhasil Dihapus.")
            elif temp:
                prevNode.next = temp
                print(f"Node Data {data} Berhasil Dihapus.")
            else:
                print("Node Tidak Ditemukan.")

    def find(self, data):
        current = self.head
        pos = 0
        while pos < self.size:
            if current.data == data:
                print(f"Position of {data} is {pos}")
                break
            else:
                current = current.next
                post += 1
        if pos == self.size:
            print("Data not found.")

    def print(self):
        text = "Linked List ="
        if self.head == None:
            print("Linked List belum terdapat data tersimpan.")
        elif self.head.next == None:
            print(text + self.head.data)
        else:
            current = self.head
            while current.next:
                text += f" {current.data} ->"
                current = current.next
            text += f" {current.data}"
            print(text)
    
data = LinkedList()
data.append(1)
data.append(2)
data.append(3)
data.append(4)
data.append(5)
# data.print()

data.insert(0)
data.print()

data.deleteNode(0)
data.print()

data.deleteNode(2)
data.print()

data.deleteNode(5)
data.print()

data.deleteNode(6)
data.print()